hazelcast/src/jepsen/hazelcast.clj

```diff
@@ -257,55 +329,235 @@
+         )
+       )
+
+     (invoke! [this test op]
+       (let [clientName (.getName conn)]
+         (try
+           (info (str " " CLIENT_ID_SEPARATOR " " clientName " " CLIENT_ID_SEPARATOR " " op))
+           (case (:f op)
+             :acquire (invoke-acquire clientName lock op)
+             :release (do
+                        (.unlock lock)
+                        (log-release-ok clientName op)
+                        (assoc op :type :ok :value {:client clientName :uid (:value op)})
+                        )
+             )
+           (catch IllegalMonitorStateException e
+             (case (:f op) :acquire (log-acquire-fail clientName op) :release (log-release-fail clientName op))
+             (assoc op :type :fail, :error :not-lock-owner)
+             )
+           (catch IOException e
+             (condp re-find (.getMessage e)
+               ; This indicates that the Hazelcast client doesn't have a remote
+               ; peer available, and that the message was never sent.
+               #"Packet is not send to owner address"
+               (do
+                 (case (:f op) :acquire (log-acquire-fail clientName op) :release (log-release-fail clientName op))
+                 (assoc op :type :fail, :error :client-down)
+                 )
+               (do
+                 (case (:f op) :acquire (log-acquire-maybe clientName op e) :release (log-release-maybe clientName op e))
+                 (assoc op :type :info, :error :io-exception)
+                 )
+               )
+             )
+           (catch Exception e
+             (case (:f op) :acquire (log-acquire-maybe clientName op e) :release (log-release-maybe clientName op e))
+             (assoc op :type :info, :error :exception)
+             )
+           )
+         )
+       )
+
+     (teardown! [this test]
+       (try
+         (.terminate (.getLifecycleService conn))
+         (catch Exception e
+           (warn "exception while client termination")
+           (.printStackTrace e)
+           )
+         )
+       )
+
+     )
+    )
+  )
+
+(defn cp-semaphore-client
+  ([] (cp-semaphore-client nil nil))
+  ([conn semaphore]
+   (reify client/Client
+     (setup! [_ test node]
+       (let [conn (connect node)
+             sem (.getSemaphore (.getCPSubsystem conn) "jepsen.cpSemaphore")
+             _ (.init sem NUMBER_OF_PERMITS)]
+         (cp-semaphore-client conn sem)
+         )
+       )
+
+     (invoke! [this test op]
+       (let [clientName (.getName conn)]
+         (try
+           (info (str " " CLIENT_ID_SEPARATOR " " clientName " " CLIENT_ID_SEPARATOR " " op))
+           (case (:f op)
+             :acquire (if (.tryAcquire semaphore 5000 TimeUnit/MILLISECONDS)
+                        (do
+                          (log-acquire-ok clientName op)
+                          (assoc op :type :ok :value {:client clientName :uid (:value op)})
```

Comment:
You use this pattern a lot in this part of the test--`(log-acquire-ok clientname op)`, followed by constructing a completion operation with `(assoc op ...)`. Have you considered having `log-acquire-ok` construct the completion operation for you? Then you could drop the do and assoc throughout these clients; might help shorten things a bit.